crates/symbolicator/src/services/shared_cache.rs

```diff
@@ -0,0 +1,346 @@
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::{anyhow, Context, Error, Result};
+use futures::TryStreamExt;
+use reqwest::{Client, StatusCode};
+use symbolic::common::ByteView;
+use tokio::fs::{self, File};
+use tokio::io::{self, AsyncWrite};
+use tokio::sync::RwLock;
+use tokio_util::io::StreamReader;
+use url::Url;
+
+use crate::cache::{
+    CacheName, FilesystemSharedCacheConfig, GcsSharedCacheConfig, SharedCacheConfig,
+};
+use crate::logging::LogError;
+use crate::services::download::measure_download_time;
+use crate::utils::gcs::{request_new_token, GcsError, GcsToken};
+
+use super::cacher::CacheKey;
+
+// TODO: get timeouts from global config?
+const CONNECT_TIMEOUT: Duration = Duration::from_secs(5);
+
+#[derive(Debug)]
+struct GcsState {
+    config: GcsSharedCacheConfig,
+    token: RwLock<Option<Arc<GcsToken>>>,
+    client: Client,
+}
+
+impl GcsState {
+    pub fn new(config: GcsSharedCacheConfig) -> Self {
+        Self {
+            config,
+            token: RwLock::new(None),
+            client: Client::new(),
+        }
+    }
+    /// Resolves a valid GCS OAuth token.
+    ///
+    /// If the cache contains a valid token, then this token is returned. Otherwise, a new token is
+    /// requested from GCS and stored in the cache.
+    async fn get_token(&self) -> Result<Arc<GcsToken>, GcsError> {
+        {
+            let guard = self.token.read().await;
+            if let Some(ref token) = *guard {
+                if !token.is_expired() {
+                    metric!(counter("shared_cache.gcs.token") += 1, "cache" => "true");
+                    return Ok(token.clone());
+                }
+            }
+        }
+
+        // Make sure only 1 token is requested at a time.
+        let mut guard = self.token.write().await;
+        let token = request_new_token(&self.client, &self.config.source_key).await?;
+        metric!(counter("shared_cache.gcs.token") += 1, "cache" => "false");
+        let token = Arc::new(token);
+        *guard = Some(token.clone());
+        Ok(token)
+    }
+
+    async fn fetch<W>(&self, key: &SharedCacheKey, writer: &mut W) -> Result<()>
+    where
+        W: tokio::io::AsyncWrite + Unpin,
```

Comment:
just curious why this `Unpin` bound is needed.

Reply:
`tokio::io::copy` requires the writer to be `Unpin`, this is used in the filesystem impl and forces us to be Unpin.